### PR TITLE
Improve lifetime errors with implicit trait object lifetimes

### PR DESCRIPTION
**Pull request by mark-i-m**
r? @matthewjasper 

cc @estebank 

I still think the ideal solution would be to construct a `BrAnon`, but that seems like a more invasive change, and can be done later. This at least gets rid of the hack in `OutliveSuggestion` and is slightly more principled.

### CODE REVIEW

File: src/librustc_mir/borrow_check/nll/region_infer/error_reporting/region_name.rs [hunk 1]
```diff
@@ -279,11 +284,30 @@ impl<'tcx> RegionInferenceContext<'tcx> {
...
+                    if free_region.bound_region.is_named() {
+                        // A named region that is actually named.
+                        Some(RegionName {
+                            name,
+                            source: RegionNameSource::NamedFreeRegion(span),
+                        })
+                    } else {
+                        // If we spuriously thought that the region is named, we should let the
+                        // system generate a true name for error messages. Currently this can
+                        // happen if we have an elided name in a trait object used in an async fn
```
Comment (matthewjasper) [inline src/librustc_mir/borrow_check/nll/region_infer/error_reporting/region_name.rs]:
```suggestion
                        // happen if we have an elided name in an async fn
```
The variant should probably be called `AnonRegionFromAsync`

File: src/librustc_mir/borrow_check/nll/region_infer/error_reporting/region_name.rs [hunk 2]
```diff
@@ -279,11 +284,30 @@ impl<'tcx> RegionInferenceContext<'tcx> {
 
             ty::ReFree(free_region) => match free_region.bound_region {
                 ty::BoundRegion::BrNamed(_, name) => {
+                    // Get the span to point to, even if we don't use the name.
                     let span = self.get_named_span(tcx, error_region, name);
```
Comment (matthewjasper) [inline src/librustc_mir/borrow_check/nll/region_infer/error_reporting/region_name.rs]: Pre-existing, but this should be `tcx.hir().span_if_local(def_id).unwrap_or(DUMMY_SP)` (where the def_id comes from the other field of `BrNamed`). The other use of `get_named_span` above should also be similarly converted.

**Review by matthewjasper — COMMENTED** [review]
This appears to apply to more than just trait objects.
r=me with comments addressed.

File: src/librustc_mir/borrow_check/nll/region_infer/error_reporting/region_name.rs (hunk 2 above)
Comment (mark-i-m) [reply src/librustc_mir/borrow_check/nll/region_infer/error_reporting/region_name.rs]: @matthewjasper `get_named_span` is a method in the `region_name` submodule, and these are its only two uses. Are you suggesting that I get rid of it altogether?

Comment (matthewjasper) [reply src/librustc_mir/borrow_check/nll/region_infer/error_reporting/region_name.rs]: Yes

**Comment by mark-i-m** [comment]
Rebased, but need to rerun tests and bless...

**Comment by mark-i-m** [comment]
@matthewjasper I think this is ready now :)

**Pull request merged**